lib/puppet/environments.rb

```diff
@@ -230,20 +233,22 @@ def create_environment(name)
       env
     end
 
-    def valid_directory?(envdir)
-      name = Puppet::FileSystem.basename_string(envdir)
-      Puppet::FileSystem.directory?(envdir) &&
-         Puppet::Node::Environment.valid_name?(name)
+    def validated_directory(envdir)
+      env_name = Puppet::FileSystem.basename_string(envdir)
+      envdir = Puppet::Environments::Directories.real_path(envdir)
+      return nil unless Puppet::FileSystem.directory?(envdir) && Puppet::Node::Environment.valid_name?(env_name)
+      return envdir
     end
 
-    def valid_directories
+    def valid_directories_by_env
```

Comment:
Nothing actually uses the actual directories from this, but I kept 'em since the previous version of the method returned those. Might clean up later though..